imap_processing/lo/l1b/lo_l1b.py

```diff
@@ -1004,3 +1045,247 @@ def create_datasets(
+    exposure_factor: np.ndarray,
+) -> xr.Dataset:
+    """
+    Calculate histogram rates by dividing reswept counts by exposure time.
+
+    For each epoch in l1b_histrates, this function finds the corresponding
+    spin interval, calculates the exposure time for 6-degree bins,
+    and divides the counts by the exposure time. The exposure time is scaled
+    by the number of ESA steps that were reswept during resweeping.
+
+    Parameters
+    ----------
+    l1b_histrates : xr.Dataset
+        The L1B histogram rates dataset containing reswept h_counts and o_counts.
+    acq_start : xr.DataArray
+        Start times for each spin cycle in MET seconds.
+    acq_end : xr.DataArray
+        End times for each spin cycle in MET seconds.
+    avg_spin_durations_per_cycle : xr.DataArray
+        Average spin duration for each cycle in seconds.
+    exposure_factor : np.ndarray
+        3D array of exposure factors (epoch, azimuth, esa_step) indicating how many
+        ESA steps were reswept during resweeping.
+
+    Returns
+    -------
+    l1b_histrates : xr.Dataset
+        Updated dataset with h_rates and o_rates added.
+    """
+    epochs = l1b_histrates["epoch"].values
+    h_counts = l1b_histrates["h_counts"].values
+    o_counts = l1b_histrates["o_counts"].values
+
+    h_rates = np.zeros_like(h_counts, dtype=float)
+    o_rates = np.zeros_like(o_counts, dtype=float)
+    num_azimuth = h_counts.shape[1]
+    exposure_times = np.zeros((len(epochs), num_azimuth, 7), dtype=float)
+
+    for epoch_idx, epoch in enumerate(epochs):
+        spin_cycle_mask = (epoch >= met_to_ttj2000ns(acq_start.values)) & (
+            epoch <= met_to_ttj2000ns(acq_end.values)
+        )
+        spin_cycle_indices = np.where(spin_cycle_mask)[0]
+
+        if len(spin_cycle_indices) == 0:
+            logger.warning(f"Epoch {epoch_idx} not found in any spin_cycle interval")
+            h_rates[epoch_idx] = np.nan
+            o_rates[epoch_idx] = np.nan
+            continue
+
+        spin_cycle_idx = spin_cycle_indices[0]
+        base_exposure_time = (
+            4 * avg_spin_durations_per_cycle.values[spin_cycle_idx] / 60
+        )
+
+        for az_idx in range(num_azimuth):
+            for esa_idx in range(7):
+                scaled_exposure = (
+                    base_exposure_time * exposure_factor[epoch_idx, az_idx, esa_idx]
+                )
+                exposure_times[epoch_idx, az_idx, esa_idx] = scaled_exposure
+
+                if scaled_exposure > 0:
+                    h_rates[epoch_idx, az_idx, esa_idx] = (
+                        h_counts[epoch_idx, az_idx, esa_idx] / scaled_exposure
+                    )
+                    o_rates[epoch_idx, az_idx, esa_idx] = (
+                        o_counts[epoch_idx, az_idx, esa_idx] / scaled_exposure
+                    )
+                else:
+                    h_rates[epoch_idx, az_idx, esa_idx] = 0
+                    o_rates[epoch_idx, az_idx, esa_idx] = 0
```

Comment:
You can get rid of the two for loops here pretty easily so we should do this.

If we don't have any exposure time do we actually want 0 rates? I would have assumed we want NaNs there since a 0 rate could be a real value. I'd suggest just ignoring the divide by zero warning and let the nan show up.
```suggestion
                scaled_exposure = (
                    base_exposure_time * exposure_factor[epoch_idx, ...]
                )
                exposure_times[epoch_idx, ...] = scaled_exposure

                with np.errstate(divide='ignore'):
                    h_rates[epoch_idx, ...] = (
                        h_counts[epoch_idx, ...] / scaled_exposure
                    )
                    o_rates[epoch_idx, ...] = (
                        o_counts[epoch_idx, ...] / scaled_exposure
                    )
```

Reply:
Good point. I'll fix this